### ISSUE TITLE
Separate import and database update scripts from website code and extend test.sh

### ISSUE THREAD

**Issue by AndrewVSutherland** [LMFDB/lmfdb#1837]
As discussed [here](https://github.com/LMFDB/lmfdb/pull/1819#issuecomment-233419972), we would like to put code that is not part of the LMFDB website into a separate directory structure and add some code validation (including pyflakes) to test.sh.

More specifically, I would propose the following:
- Move the directory lmfdb/import_scripts up one level so that import_scripts is parallel to lmfdb.  Inside import_scripts is a directory tree that effectively parallels the directory tree under lmfdb (we might want to shorten and/or generalize the name "import_scripts" to make our lives simpler).
- Move import/update scripts that currently reside in directories under lmfdb (e.g. those in lmfdb/hilbert_modular_forms, for example) into the parallel directory under import_scripts (e.g. import_scripts/hilbert_modular_forms).
- Extend test.sh to run pyflakes on the entire lmfdb/... subtree (pyflakes is fast enough for this to be quite feasible, it only takes a few seconds).
- Extend test.sh to search for uses of eval and sage_eval in all .py code in the lmfdb/... subtree.
- I think test scripts (e.g. test_....py) should stay in the lmfdb/... subtree even though they are not part of the website.

Please feel free to suggest modifications and additions to this proposal.  I have tagged this for Release 1.1.

**Comment by JohnCremona** [comment]
I suggest that the very first step should be to action your first bullet point, i.e. just move the import_scripts directory.  I know that the import scripts I have (which are not in there) will not work after moving, without a small amount of editing, and the same might be true for the scripts already in in the separate directory.  So the question then is: does one of us try to make those edits at the same time, or do we recruit the "owners" of the 8 subdirectories  import_scripts/* to deal with their own?

**Comment by edgarcosta** [comment]
One could use sys.path.append as a quick and dirty fix.

**Comment by jwj61** [comment]
It is probably a combination of sys.path.append so scripts will be able to import lmfdb code and code to find password files.

It looks like the owners of the current import_scripts directories are me, Samuele, and maybe Stefan for Maass.  There is also code from Paul which is no longer used.

**Comment by JohnCremona** [comment]
OK, so let's move them and ping them (Samuele and Stefan/Stephan?) on the pull request.

**Comment by JohnCremona** [comment]
I'll do this now and make a PR.

**Comment by JohnCremona** [comment]
OK, I did a simple move -- no editing of files yet -- and pushed the result to the branch move_import_scripts in *this* repository.   If people want to also edit the files to see if things need changing to make them work they can either push directly to this branch (if they have permission) or make a PR to it otherwise.  @lemurell @sanni85

**Comment by sanni85** [comment]
No problem in moving the folder, the scripts will not work, but I know how to fix them.
I am still in Sarajevo, so I will do something, if needed, next week.

By the way please do not run pyflakes on rep_galois_modl : it is very much in progress (wrote in 4h at AIM) and so most of the code is just a copy from the lattice folder.

**Comment by AndrewVSutherland** [comment]
@sanni85 pyflakes has already been run on rep_galois_modl and it runs cleanly, as it now does on all the code in the LMFDB/lmfdb project. The goal is to keep the LMFDB codebase in a state where pyflakes runs cleanly; if you have code that is in a state where pyflakes won't run on it (the moral equivalent of C code that will not compile), better to keep it in your own fork.

**Comment by JohnCremona** [comment]
After three months we have all forgotten about this, I think.  The original branch I made no longer merges onto master, so I will remake it and do what I originall said I would and actually make a pull request.  This will not affect the website (that's the point).

**Comment by AndrewVSutherland** [comment]
Great, thanks!

**Comment by JohnCremona** [comment]
If someone would make a list of other scripts in ./lmfdb that would help -- or just find them all and move them.  There are more than I thought.  In most cases just moving the files (using git mv, after creating a new subdirectory of ./scripts) is enough.  Perhaps it would be OK to just do it -- the script "owners" can fix anything needed afterwards?

**Comment by JohnCremona** [comment]
@roed314 I did not spot that in the recent big merge we did for abvar that you put a couple of files into the otherwise empty import_scripts directory.  Can you move lmfdb/import_scripts/abvar to scripts/abvar please?  The directory lmfdb/import_scripts is now obsolete.

**Comment by JohnCremona** [comment]
Perhaps this should a separate issue, but I just remembered that one can run "./test.sh coverage" and get a report on which lines of code are covered by the tests.  Overall we are at 64% now, but some files are very low.  Perhaps we should get someone to try to increase that?  I also thought that it wouldbe nice if the test output could show the time for each test and not just "ok".  nosetest does have a profiling option -- you can add --with-profile to the command line inside test.sh and it produces a lot of information only a small part of which is relevant.  This might be a nice job for someone to do?

**Comment by edgarcosta** [comment]
I think those numbers should be advertised on an automatic basis.
For example, using https://codecov.io  (see https://github.com/codecov/example-python).
Let's first get automated testing running.

Regarding the timing, in travis, I'm already testing each folder/file individually, so you can already get a good idea where most of the time is spent, see:
https://travis-ci.org/edgarcosta/lmfdb/builds/273341808

Not the line between OK and  INFO, has nothing to with what follows INPUT. (it just a line reporting that command started to be run on the background)

**Comment by davidlowryduda** [comment]
Does travis also allow for coverage? Or is using something like codecov necessary?

**Comment by edgarcosta** [comment]
Nop. 

We would use travis to upload the coverage xml to codecov.io.

**Comment by edgarcosta** [comment]
I'm trying it out to see how it works.

**Comment by JohnCremona** [comment]
As I said earlier on some other thread, our test script can report on
coverage already.

**Comment by edgarcosta** [comment]
Yes, it does.

The point is to build up on that, and upload those reports automatically, so we can track our coverage in automatic way, instead of having to ask people to run ./test.sh with coverage.

@davidlowryduda  I was wrong, travis-ci claims that it "integrates" coveralls (an alternative to codecov). But I don't it integrates it any better than codecov for our purposes. Also, check this out:
https://github.com/codecov/browser-extension#codecov-extension

**Comment by edgarcosta** [comment]
@JohnCremona and @davidlowryduda check this:
https://codecov.io/gh/edgarcosta/lmfdb/branch/travis_wcoverage

**Comment by edgarcosta** [comment]
The coverage is not yet correct (something to due with running test in parallel, I'm fixing it)

**Comment by edgarcosta** [comment]
I have fixed it. The different  percentage numbers have to do how both softwares decide to compute the coverage percentage, but the number of misses and partial hits it is the same.

**Comment by edgarcosta** [comment]
I believe that now we should advertise: https://codecov.io/gh/LMFDB/lmfdb
on the README.
What do you think about that?

**Comment by edgarcosta** [comment]
Also, if you find codecov to spammy, we can tweak its comments. See: https://github.com/codecov/support/wiki/Pull-Request-Comments

**Comment by AndrewVSutherland** [comment]
I believe all the items raised in this issue have now been addressed!  Of course there is always more we could do, but I think this issue can now be safely closed.
